### ISSUE TITLE
[BUG] Virtualenv does not get updated when already exists

### ISSUE THREAD

**Issue by sxooler** [mlflow/mlflow#10525]
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.1.1
- Tracking server: 2.1.1

### System information

- **OS Platform and Distribution**: Ubuntu 20.04 (WSL2)
- **Python version**: 3.10.13

### Describe the problem

I am running my training scripts using `mlflow run` (locally) against project in gitlab (something like `mlflow run git@gitlab.com:custom-repo.git#scripts/pv_model -e main -P parameter1=test --experiment-name test-model`). I am using `MLProject` and `python_env.yaml`. When I run the same command again and the project was updated in meantime, the previous environment is used and dependencies are not updated resulting in failed run.

Is that expected behavior? I can always clean the created virtualenv and force mlflow to create it again and run the installation of all dependencies. But it was already confusing for me (especially in the beginning) when I was not sure why my pipeline is failing and then I realized it's because the previous venv is reused and not updated.

Note: neither `python_env.yaml` nor `MLproject` is changed between the runs. The content of `python_evn.yaml` looks like:
```
python: "3.10"
build_dependencies:
  - pip
dependencies:
  - mlflow==2.1.1
  - custom-repo
```
I am expecting it will take the latest version of `custom-repo`. If there was `1.1` as the latest version of `custom-repo` then it will be installed locally to virtualenv. In another run when the version of `custom-repo` was changed to `1.2`, it will not get installed as the previous virtualenv (with `1.1` version) will be used without any update (causing the issues described above).

### Code to reproduce issue

I do not have any way how to reproduce it as the first run is successful but next runs might fail when the dependencies changes (breaking change).

### Stack trace

There is not specific error, during rerunning, previously created environment is used which is then causing various issues depending on the project:
```
2023/11/28 13:34:44 INFO mlflow.utils.virtualenv: Installing python 3.10.13 if it does not exist
2023/11/28 13:34:44 INFO mlflow.utils.virtualenv: Environment /home/user1/.mlflow/envs/mlflow-ad68ca8dcbe8c9697931caf3c6e85ba939168ae3 already exists
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [X] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

**Comment by B-Step62** [comment]
@sxooler Yes, either having a new command or a flag to update virtualenv makes sense to me, which to take is more like personal preference imo:) Would you mind taking this issue and filing a PR?

Also totally agree that the behavior should be more visible to new users. We should probably update docs to explicitly mention this.

**Comment by sxooler** [comment]
@B-Step62 Could you point me in the right direction? I guess this makes sense only for a local runs, therefore `projects/backend/local.py` should be mainly affected. Or does it make sense to implement sth similar elsewhere?
